File: library/kotlin/io/envoyproxy/envoymobile/stats/DistributionImpl.kt
package io.envoyproxy.envoymobile

import io.envoyproxy.envoymobile.engine.EnvoyEngine
import java.lang.ref.WeakReference

/**
 * Envoy implementation of a `Distribution` for measurements of quantile data for int values
 */
internal class DistributionImpl : Distribution {
  var envoyEngine: WeakReference<EnvoyEngine>
  var series: String
  var tags: Map<String, String>

  internal constructor(engine: EnvoyEngine, elements: List<Element>, tags: List<Tag>) {
    this.envoyEngine = WeakReference<EnvoyEngine>(engine)
    this.series = elements.joinToString(separator = ".") { it.value }
    this.tags = convert(tags)
  }

  private constructor(engine: WeakReference<EnvoyEngine>, series: String, tags: List<Tag>) {
    this.envoyEngine = engine
    this.series = series
    this.tags = convert(tags)
  }

  override fun recordValue(value: Int) {
    envoyEngine.get()?.recordHistogramValue(series, tags, value)
  }

  override fun attach(tags: List<Tag>): Distribution {
    return DistributionImpl(envoyEngine, series, tags)
  }
}

File: library/swift/stats/DistributionImpl.swift
@_implementationOnly import EnvoyEngine
import Foundation

/// The implementation of distribution tracking quantile/sum/average stats
@objcMembers
final class DistributionImpl: NSObject, Distribution {
  private let series: String
  private let tags: [String: String]
  private weak var engine: EnvoyEngine?

  init(elements: [Element], tags: [Tag], engine: EnvoyEngine) {
    self.series = elements.map { $0.value }.joined(separator: ".")
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  init(series: String, tags: [Tag], engine: EnvoyEngine?) {
    self.series = series
    self.engine = engine
    self.tags = Stats.convert(tags: tags)
    super.init()
  }

  /// Record a new int value for the distribution.
  func recordValue(value: Int) {
    // TODO(jingwei99) potentially surface error up if engine is nil.
    self.engine?.recordHistogramValue(self.series, tags: tags, value: numericCast(value))
  }

  func attach(tags: [Tag]) -> Distribution {
    return DistributionImpl(series: self.series, tags: tags, engine: self.engine)
  }
}
